File: runtime/tests/integration/timers_tests.rs
use std::time::Instant;

use assert_fs::prelude::*;
use zinnia_runtime::resolve_path;
use zinnia_runtime::run_js_module;

#[tokio::test]
async fn set_timeout_works() -> Result<(), Box<dyn std::error::Error>> {
  let mod_js = assert_fs::NamedTempFile::new("timers_tests.js")?;
  mod_js.write_str(
    r#"
setTimeout(() => {}, 50);
"#,
  )?;

  let now = Instant::now();
  run_js_module(
    &resolve_path(&mod_js.path().to_string_lossy())?,
    &Default::default(),
  )
  .await?;
  let elapsed = now.elapsed().as_millis();

  assert!(
    elapsed > 40 && elapsed < 100,
    "setTimeout(50) should take between 40 to 100 ms to execute, but took {elapsed} ms instead",
  );

  Ok(())
}

File: runtime/tests/js/timers_tests.js
const start = Date.now();

await new Promise((resolve) => {
  setTimeout(resolve, 50);
});

const duration = Date.now() - start;
const min = 40;
const max = 120;

if (duration < min || duration > max) {
  throw new Error(
    `setTimeout(50) should take between ${min} to ${max} ms to execute, but took ${duration} ms instead`,
  );
}
